rclpy/actions/minimal_action_server/examples_rclpy_minimal_action_server/server.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from example_interfaces.action import Fibonacci
+
+import rclpy
+from rclpy.action import ActionServer, CancelResponse, GoalResponse
+from rclpy.callback_groups import ReentrantCallbackGroup
+from rclpy.executors import MultiThreadedExecutor
+from rclpy.node import Node
+
+
+class MinimalActionServer(Node):
+
+    def __init__(self):
+        super().__init__('minimal_action_server')
+
+        self._action_server = ActionServer(
+            self,
+            Fibonacci,
+            'fibonacci',
+            execute_callback=self.execute_callback,
+            callback_group=ReentrantCallbackGroup(),
+            goal_callback=self.goal_callback,
+            cancel_callback=self.cancel_callback)
+
+    def destroy(self):
+        self._action_server.destroy()
+        super().destroy_node()
+
+    def goal_callback(self, goal_request):
+        """Accepts or rejects a client request to begin an action."""
+        # This server allows multiple goals in parallel
+        self.get_logger().info('Received goal request')
+        return GoalResponse.ACCEPT
+
+    def cancel_callback(self, goal_handle):
+        """Accepts or rejects a client request to cancel an action."""
+        self.get_logger().info('Received cancel request')
+        return CancelResponse.ACCEPT
+
+    async def execute_callback(self, goal_handle):
+        """Executes a goal."""
+        self.get_logger().info('Executing goal...')
+
+        # Append the seeds for the Fibonacci sequence
+        feedback_msg = Fibonacci.Feedback()
+        feedback_msg.sequence = [0, 1]
+
+        # Start executing the action
+        for i in range(1, goal_handle.request.order):
+            if goal_handle.is_cancel_requested:
+                goal_handle.set_canceled()
+                self.get_logger().info('Goal canceled')
+                return Fibonacci.Result()
+
+            # Update Fibonacci sequence
+            feedback_msg.sequence.append(feedback_msg.sequence[i] + feedback_msg.sequence[i-1])
+
+            self.get_logger().info('Publishing feedback: {0}'.format(feedback_msg.sequence))
+
+            # Publish the feedback
+            goal_handle.publish_feedback(feedback_msg)
+
+            # Sleep for demonstration purposes
+            time.sleep(1)
```

Comment:
nit, should be able to use asyncio sleep to let the executor do other stuff (although, there's no other stuff being done in this example)
```python
await asyncio.sleep(1)
```

Reply:
Good point, but there must be a bug in `rclpy` because changing this to `await asyncio.sleep(1)` causes it to hang forever at this line. If I also change the executor to a `SingleThreadedExecutor`, then there's an assertion error with the following trace:

```
Traceback (most recent call last):
  File "/home/jacob/ws/actions_ws/install/examples_rclpy_minimal_action_server/lib/examples_rclpy_minimal_action_server/server", line 11, in <module>
    load_entry_point('examples-rclpy-minimal-action-server==0.6.1', 'console_scripts', 'server')()
  File "/home/jacob/ws/actions_ws/install/examples_rclpy_minimal_action_server/lib/python3.5/site-packages/examples_rclpy_minimal_action_server/server.py", line 103, in main
    rclpy.spin(minimal_action_server, executor=executor)
  File "/home/jacob/ws/actions_ws/install/rclpy/lib/python3.5/site-packages/rclpy/__init__.py", line 119, in spin
    executor.spin_once()
  File "/home/jacob/ws/actions_ws/install/rclpy/lib/python3.5/site-packages/rclpy/executors.py", line 573, in spin_once
    raise handler.exception()
  File "/home/jacob/ws/actions_ws/install/rclpy/lib/python3.5/site-packages/rclpy/task.py", line 207, in __call__
    self._handler.send(None)
  File "/home/jacob/ws/actions_ws/install/rclpy/lib/python3.5/site-packages/rclpy/action/server.py", line 323, in _execute_goal
    execute_result = await await_or_execute(execute_callback, goal_handle)
  File "/home/jacob/ws/actions_ws/install/rclpy/lib/python3.5/site-packages/rclpy/executors.py", line 89, in await_or_execute
    return await callback(*args)
  File "/home/jacob/ws/actions_ws/install/examples_rclpy_minimal_action_server/lib/python3.5/site-packages/examples_rclpy_minimal_action_server/server.py", line 80, in execute_callback
    await asyncio.sleep(1)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 516, in sleep
    return (yield from future)
  File "/usr/lib/python3.5/asyncio/futures.py", line 362, in __iter__
    assert self.done(), "yield from wasn't used with future"
AssertionError: yield from wasn't used with future
```

I'll see if I can figure out what's gone wrong.

Reply:
This looks like me being wrong about what `asyncio.sleep()` expects. https://github.com/python/cpython/blob/6b0d50d92e80faaed8a043e6554ccb5d046114a5/Lib/asyncio/tasks.py#L517-L532

Looks like it needs an asyncio event loop. You can ignore my comments about using it

Reply:
I see. 
I can reproduce the same issue with a Service server as well. It seems like something rclpy should be able to handle though, so I've opened an issue ros2/rclpy#279.